deterministic_kernel.hpp

```diff
@@ -0,0 +1,247 @@
+ * The code contained herein is licensed under the GNU General Public
+ * License. You may obtain a copy of the GNU General Public License
+ * Version 2 or later at the following locations:
+ *
+ * http://www.opensource.org/licenses/gpl-license.html
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+#ifndef POPS_DETERMINISTIC_KERNEL_HPP
+#define POPS_DETERMINISTIC_KERNEL_HPP
+
+#include <vector>
+#include <cmath>
+#include <tuple>
+
+#include "raster.hpp"
+
+#ifndef M_PI
+#define M_PI 3.14159265358979323846
+#endif
+#ifndef PI
+#define PI M_PI
+#endif
+
+namespace pops {
+
+using std::pow;
+using std::tan;
+using std::atan;
+using std::exp;
+using std::log;
+using std::ceil;
+using std::abs;
+
+/*!
+ * Cauchy distribution
+ * Includes probability density function and cumulative distribution function
+ * pdf returns the probability that the variate has the value x
+ * cdf returns the upper range that encompasses x percent of the distribution (e.g for 99% input .99)
+ */
+class CauchyDistribution
+{
+public:
+    CauchyDistribution(double scale, double locator)
+        : s(scale), t(locator)
+    {}
+
+    double pdf(double x)
+    {
+        return 1 / ((s*PI)*(1 + (pow((x - t)/s, 2))));
+
+    }
+
+    double cdf(double x)
+    {
+        if ( s == 1 && t == 0) {
+            // checking upper half so multiply result by 2
+            return tan((x / 2) * M_PI) * 2;
+        } else {
+            return (tan((x/2)*M_PI + (atan(-t/s))) * s + t) * 2;
+        }
+    }
+private:
+    // scale parameter - 1 for standard
+    double s;
+    // location parameter - location of peak - 0 for standard
+    double t;
+};
+
+/*!
+ * Exponential distribution
+ * Includes probability density function and cumulative distribution function
+ * pdf returns the probability that the variate has the value x
+ * cdf returns the upper range that encompasses x percent of the distribution (e.g for 99% input 0.99)
+ */
+class ExponentialDistribution
+{
+public:
+ExponentialDistribution(double scale)
+        : beta(scale)
+    {}
+    // assumes mu is 0 which is traditionally accepted
+    double pdf(double x)
+    {
+        return (1/beta)*(exp(-x/beta));
+    }
+
+    double cdf(double x)
+    {
+        if ( beta == 1) {
+            return -log(1-x);
+        } else {
+            return -beta * log(1-x);
+        }
+    }
+private:
+    // scale parameter - 1 for standard
+    // equal to 1/lambda
+    double beta;
+};
+
+/*!
+ * Dispersal kernel for deterministic spread to cell with highest probability of spread
+ *
+ * Dispersal Kernel type determines use of Exponential or Cauchy distribution
+ * to find probability.
+ *
+ * dispersal_percentage is the percent of all possible dispersal to be included
+ * in the moving window size (e.g for 99% input 0.99).
+ *
+ * Useful for testing as it is deterministic and provides fully replicable results
+ */
+template<typename IntegerRaster>
+class DeterministicDispersalKernel
+{
+protected:
+    IntegerRaster dispersers_;
```

Comment:
Since IntegerRaster dispersers is passed into the RadialDispersalKernel constructor, I also changed that to be const, but if this is wrong I can change it back.